src/iceberg/update/update_partition_spec.cc

```diff
@@ -0,0 +1,488 @@
+                        "Cannot find partition field to rename: {}", name);
+
+  // Cannot delete and rename
+  ICEBERG_BUILDER_CHECK(!deletes_.contains(field_it->second->field_id()),
+                        "Cannot delete and rename partition field: {}", name);
+
+  renames_[name] = new_name;
+  return *this;
+}
+
+Result<UpdatePartitionSpec::ApplyResult> UpdatePartitionSpec::Apply() {
+  ICEBERG_RETURN_UNEXPECTED(CheckErrors());
+
+  // Reserve capacity to avoid reallocations
+  const size_t existing_fields_count = spec_->fields().size();
+  const size_t adds_count = adds_.size();
+  std::vector<PartitionField> new_fields;
+  new_fields.reserve(existing_fields_count + adds_count);
+
+  // Process existing fields
+  for (const auto& field : spec_->fields()) {
+    if (!deletes_.contains(field.field_id())) {
+      // Field is kept, check for rename
+      auto rename_it = renames_.find(std::string(field.name()));
+      if (rename_it != renames_.end()) {
+        new_fields.emplace_back(field.source_id(), field.field_id(), rename_it->second,
+                                field.transform());
+      } else {
+        new_fields.push_back(field);
+      }
+    } else if (format_version_ < 2) {
+      // In V1, deleted fields are replaced with void transform
+      auto rename_it = renames_.find(std::string(field.name()));
+      std::string field_name =
+          rename_it != renames_.end() ? rename_it->second : std::string(field.name());
+      new_fields.emplace_back(field.source_id(), field.field_id(), field_name,
+                              Transform::Void());
+    }
+    // In V2, deleted fields are simply removed
+  }
+
+  // Add new fields
+  new_fields.insert(new_fields.end(), adds_.begin(), adds_.end());
+
+  // Determine the new spec ID
+  int32_t new_spec_id = spec_ ? spec_->spec_id() + 1 : PartitionSpec::kInitialSpecId;
+
+  // In V2, if all fields are removed, reset last_assigned_partition_id to allow
+  // field IDs to restart from 1000 when fields are added again
+  int32_t last_assigned_id = last_assigned_partition_id_;
+  if (format_version_ >= 2 && new_fields.empty()) {
+    last_assigned_id = PartitionSpec::kLegacyPartitionDataIdStart - 1;
+  }
+
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto new_spec,
+      PartitionSpec::Make(*schema_, new_spec_id, std::move(new_fields),
+                          /*allow_missing_fields=*/false, last_assigned_id));
+
+  ApplyResult result;
+  auto spec_id = new_spec->spec_id();
+
+  // Add the new partition spec
+  result.updates.emplace_back(
+      std::make_unique<table::AddPartitionSpec>(std::move(new_spec)));
+
+  // If set_as_default_ is true, set this spec as the default
+  if (set_as_default_) {
+    result.updates.emplace_back(
+        std::make_unique<table::SetDefaultPartitionSpec>(spec_id));
+  }
+
+  return result;
+}
+
+int32_t UpdatePartitionSpec::AssignFieldId() { return ++last_assigned_partition_id_; }
+
+PartitionField UpdatePartitionSpec::RecycleOrCreatePartitionField(
```

Comment:
Please add the comment from the java impl of `recycleOrCreatePartitionField`. I think it is helpful for understanding it.